readme.md
# Platinum Renegade Wiki

This Readme should help you to edit and deploy your own version of this Wiki

## Setup

1.  Clone this repo
2.  Install python
3.  Install requirements with `pip install -r requirements.txt`

## Local Development

In the main directory of this repo run:
```
python -m mkdocs serve
```
## Fix table layout

This script fixes the table formatting and removes unused columns.
Example to run it on everything.
```sh
python fix_table_formatting.py ./docs/*.md ./docs/*/*.md
```

## Fix links

This script fixes all the link locations in a range of files.
Example to fix the links for all pokemon changes.
If you have new links that need to be added add them to `links.txt`
```sh
python fix_links.py /pokemon_changes/*.md
```

## PokeAPI related scripts

The `download_pokemon.py` script loads a lot of data from the PokeAPI and
caches it in `temp`.

### Create Level-Up tables

The `create_level_up_table.py` script was used to add move infos to all level up moves. The move-info is mostly grabbed from PokeAPI but contains the changes done by Drayano.

### Create Stat tables

The `create_stat_table.py` script was used to add stats to all pokemon regardless of change. The stats is mostly grabbed from PokeAPI but contains the changes done by Drayano.

### Create Encounter tables

The `create_encounter_table.py` script was used to add encounter locations to all Pokemons. Thanks to @Paco-m-c for the suggestion.

### Create Ability tables

The `create_ability_table.py` script was used to add the ability to all pokemon regardless of change. The stats is mostly grabbed from PokeAPI but contains the changes done by Drayano.

docs/changelog.md
# Changelog

## v1.3.0 - 16/04/2019

#### Breaking Changes:
* The patching system for the hack has changed.
- There's now a base patch to change your Platinum into Renegade Platinum, one each for the 3541 numbered ROM and the 4997 numbered ROM.
- You can then apply smaller patches on top of Renegade Platinum, including the classic mode, more normal shiny rates and a speed up patch to make the game faster.

For details on applying these smaller patches, please see the README included in the folder with them.
The base Renegade Platinum that you get after the initial patch uses what was previously the Complete and Shiny Boosted version (i.e. all Pokémon changes are present and the shiny rate is 1/512).

#### New Features:
- You can now battle Steven when talking to him in Oreburgh City if you've defeated the Pokémon League.
- A Hidden Power teller NPC has been placed in the Jubilife Trainers' School. He can tell you the type of your Pokémon's Hidden Power (but not the power, sadly).
- The obedience check for traded Pokémon in battles has been disabled. This means all Pokémon - including those from the trades in the game - should listen to you at all times. Relevant dialogue has been modified to reflect this.
- Pokémon now have their encounter locations listed accurately inside the Pokédex. (Big thanks to Mikelan98 for this!)

#### Pokémon changes (both versions):
- Spheal now evolves into Sealeo at Level 24.
- Sealeo now evolves into Walrein at Level 40.
- Aron now evolves into Lairon at Level 24.
- Lairon now evolves into Aggron at Level 40.
- Sealeo's level up move levels have been adjusted.
- Walrein's level up move levels have been adjusted.
- Lairon's level up move levels have been adjusted.
- Aggron's level up move levels have been adjusted.
- Glalie can now learn Rock Tomb by TM.
- Zigzagoon, Linoone, Gligar and Gliscor can now learn Rock Climb by HM (they have access to the move in Gen V onwards).
- Oddish, Gloom and Vileplume have had their level up moves reshuffled and Oddish and Gloom now learn Sludge by level.

#### Changes:
- The man in Celestic Town that gives you various "glasses" items now gives you all items when spoken to instead of an item depending on the time.
- The boy who talks to you as you enter Oreburgh City now gives you an Oval Stone (helpful if you get Happiny from the Egg you receive in the school).
- The item changes document now lists the location of key items.

#### Bug Fixes:
- Modified Maylene's dialogue when entering Route 217 to avoid the text getting cut off if the player's name was long.
- The rock that appears on Fullmoon Island now uses the Moss Rock overworld, preventing the player from destroying it by using Rock Smash from the menu.
- Fixed an error where the encounters for fishing in the wild document didn't match what was in game.
- Fixed an error in the wild document listing Floatzel as a 30% encounter at night on Route 213.
- Fixed a couple Trainers that had "nd" in their name instead of an ampersand (&).
- Fixed a couple Trainers who didn't have their rematch teams properly edited.
- Fixed a bug where one of Lucian's Pokémon was the incorrect level.
- Fixed a bug where some Trainers in Victory Road had a Dusknoir and Banette incorrectly knowing Shadow Force.

## v1.2.1 - 18/01/2019

#### Pokémon changes (both versions):
- Glalie can now learn Rock Polish by TM.

#### Changes:
- The NPC who fixes saves from the earlier versions now doesn't affect the Celebi event. Instead, you should now be able to interact with the shrine in Celestic Town again after updating to v1.2.1 and then subsequently get the Celebi event working. This does mean that people who already caught Celebi can get a second Celebi!

#### Bug Fixes:
- Fixed a bug where the Celebi event failed to activate. If you previously collected the GS Ball from the shrine, the event should work now.

## v1.2.0 - 14/01/2019

#### New Features:
- A scientist in the Berry Master's house on Route 208 now sells you Berries normally unobtainable in Platinum (e.g. Liechi, Salac).
- The Moss Rock and the Ice Rock have returned. While they no longer have an effect on Eevee, they will give you a Leaf Stone and an Ice Stone respectively.
- Added a "Type Changes" document that lists a justification for each type change as well as the types that are changed for any Pokémon.
- Added a "Frequently Asked Questions" document that has answers to some of the more common questions received.

#### Pokémon changes (both versions):
- Chikorita, Bayleef and Meganium now learn Draining Kiss and Moonblast by level.
- Totodile, Croconaw and Feraligatr now learn Night Slash by level.
- Milotic now learns Moonblast by level.
- Glalie is now able to learn Rock Slide by level and TM, and Stone Edge, Rock Smash and Rock Climb by TM/HM.

#### Pokémon changes (Complete version only):
- Meganium is now Grass/Fairy type.
- Feraligatr is now Water/Dark type.
- Milotic is now Water/Fairy type.
- Glalie is now Ice/Rock type.

#### Bug Fixes:
- Fixed a bug where Ghost-type attacks were neutral effectiveness against Dark-type Pokémon.
- Fixed a bug where Ghost-type attacks were not very effective against Dragon-type Pokémon.
- Fixed a bug where the NPC that upgrades your save was sometimes making the Celebi event not work.
- Fixed a bug where the Deoxys event wouldn't work by modifying the event slightly. This should now work on save files where it didn't before (if the League has been beaten).
- Fixed a bug where a pair of Trainers in Wayward Cave didn't see far enough to force a double battle.
- Fixed a bug where Volkner incorrectly used two normal Rotom in his Battleground rematch instead of the Heat and Wash forms.
- Fixed a bug where you could battle the Snowpoint Temple guards with only one Pokémon, causing you to send out a glitch.
- Fixed a bug where an NPC in the Veilstone Department Store made a reference to giving you the Counter Pokétch app, which never happened in Renegade Platinum.

script copy 2.py
import glob
import requests
import json
import numpy as np
np.set_printoptions(precision=3)

def add_type():
    pass

def get_type(pokemon_id):
    url = "https://pokeapi.co/api/v2/pokemon/{}/".format(int(pokemon_id))
    pokeapi = json.loads(requests.get(url, allow_redirects=True).content)

    if len(pokeapi["types"]) == 1:
        type_ = [pokeapi["types"][0]["type"]["name"]]
        return type_
    else:
        type_ = [pokeapi["types"][0]["type"]["name"], pokeapi["types"][1]["type"]["name"]]
        return type_


def get_effectiveness(type_a, type_b = None):
    type_to_id = {
        "normal" : 1,
        "fighting" : 2,
        "flying" : 3,
        "poison" : 4,
        "ground": 5,
        "rock": 6,
        "bug": 7,
        "ghost": 8,
        "steel": 9,
        "fire": 10,
        "water": 11,
        "grass": 12,
        "electric": 13,
        "psychic": 14,
        "ice": 15,
        "dragon": 16,
        "dark": 17,
        "fairy": 18
    }

    id_to_type = [
        "normal",
        "fighting",
        "flying",
        "poison",
        "ground",
        "rock",
        "bug",
        "ghost",
        "steel",
        "fire",
        "water",
        "grass",
        "electric",
        "psychic",
        "ice",
        "dragon",
        "dark",
        "fairy"
        ]

    e_m =np.array([#normal fighting flying poison ground rock  bug ghost steel fire water grass electric psychic  ice dragon dark fairy
                    [    1,     2.0,     1,     1,     1,   1,   1,  0.0,    1,   1,    1,    1,       1,      1,   1,     1,   1,    1], #normal
                    [    1,       1,   2.0,     1,     1, 0.5, 0.5,    1,    1,   1,    1,    1,       1,    2.0,   1,     1, 0.5,  2.0], #fighting
                    [    1,     0.5,     1,     1,   0.0, 2.0, 0.5,    1,    1,   1,    1,  0.5,     2.0,      1, 2.0,     1,   1,    1], #flying
                    [    1,     0.5,     1,   0.5,   2.0,   1, 0.5,    1,    1,   1,    1,  0.5,       1,    2.0,   1,     1,   1,  0.5], #poison
                    [    1,       1,     1,   0.5,     1, 0.5,   1,    1,    1,   1,  2.0,  2.0,       0,      1, 2.0,     1,   1,    1], #ground
                    [  0.5,     2.0,   0.5,   0.5,   2.0,   1,   1,    1,  2.0, 0.5,  2.0,  2.0,       1,      1,   1,     1,   1,    1], #rock
                    [    1,     0.5,   2.0,     1,   0.5, 2.0,   1,    1,    1, 2.0,    1,  0.5,       1,      1,   1,     1,   1,    1], #bug
                    [    0,       0,     1,   0.5,     1,   1, 0.5,  2.0,    1,   1,    1,    1,       1,      1,   1,     1, 2.0,    1], #ghost
                    [  0.5,     2.0,   0.5,     0,   2.0, 0.5, 0.5,    1,  0.5, 2.0,    1,  0.5,       1,    0.5, 0.5,   0.5,   1,  0.5], #steel
                    [    1,       1,     1,     1,   2.0, 2.0, 0.5,    1,  0.5, 0.5,  2.0,  0.5,       1,      1, 0.5,     1,   1,  0.5], #fire
                    [    1,       1,     1,     1,     1,   1,   1,    1,  0.5, 0.5,  0.5,  2.0,     2.0,      1, 0.5,     1,   1,    1], #water
                    [    1,       1,   2.0,   2.0,   0.5,   1, 2.0,    1,    1, 2.0,  0.5,  0.5,     0.5,      1, 2.0,     1,   1,    1], #grass
                    [    1,       1,   0.5,     1,   2.0,   1,   1,    1,  0.5,   1,    1,    1,     0.5,      1,   1,     1,   1,    1], #electric
                    [    1,     0.5,     1,     1,     1,   1, 2.0,  2.0,    1,   1,    1,    1,       1,    0.5,   1,     1, 2.0,    1], #psychic
                    [    1,     2.0,     1,     1,     1, 2.0,   1,    1,  2.0, 2.0,    1,    1,       1,      1, 0.5,     1,   1,    1], #ice
                    [    1,       1,     1,     1,     1,   1,   1,    1,    1, 0.5,  0.5,  0.5,     0.5,      1, 2.0,   2.0,   1,  2.0], #dragon
                    [    1,     2.0,     1,     1,     1,   1, 2.0,  0.5,    1,   1,    1,    1,       1,      0,   1,     1, 0.5,  2.0], #dark
                    [    1,     0.5,     1,   2.0,     1,   1, 0.5,    1,  2.0,   1,    1,    1,       1,      1,   1,     0, 0.5,    1] #fairy
                ])

    a_id = type_to_id[type_a]
    a_e = e_m[a_id-1, :]
    if not (type_b is None):
        b_id = type_to_id[type_b]
        b_e = e_m[b_id-1, :]
    else:
        b_e = np.ones(len(e_m[0, :]))

    e = a_e * b_e

    eff = {0:[],0.25:[],0.5:[],1:[],2:[],4:[]}
    for i in range(len(e)):
        eff[e[i]].append(id_to_type[i])

    return eff

def get_type_lines(t):
    if len(t) == 2:
        eff = get_effectiveness(t[0], t[1])
    else:
        eff = get_effectiveness(t[0])

    type_lines = []
    type_lines.append("## Type\n")
    type_lines.append("\n")
    if len(t) == 1:
        type_lines.append("![][{}]\n".format(t[0]))
    else:
        type_lines.append("![][{}]  ![][{}]\n".format(t[0],t[1]))

    type_lines.append("\n")
    type_lines.append("### Defenses\n")
    type_lines.append("\n")

    type_lines.append("Immune ×0 | Resistant ×¼ | Resistant ×½ | Normal ×1 | Weak ×2 | Weak ×4\n")
    type_lines.append("---       | ---          | ---          | ---       | ---     | ---\n")
    eff_line = ""
    for i in [0,0.25,0.5,1.0,2.0,4.0]:
        if len(eff[i]) == 0:
             eff_line += "&nbsp;"
        else:
            for t in eff[i]:
                eff_line += "![][{}]<br>".format(t)
        eff_line += "| "

    type_lines.append(eff_line+"\n")

    type_lines.append("\n")

    return type_lines

if __name__ == "__main__":

    for f in glob.glob("docs/pokemon_changes/*.md"):
        file_lines = open(f, "r", encoding="utf-8").readlines()

        skip = False
        for line in file_lines:
            if line.startswith("## Type"):
                skip = True
                break
        if not skip:

            poke_id = file_lines[0].replace("# ", "").split("-")[0].strip()

            t = get_type(poke_id)

            type_lines = get_type_lines(t)

            insert_line = 0
            for i in range(len(file_lines)):
                if file_lines[i].startswith("## "):
                    insert_line = i
                    break
            output = file_lines[:insert_line]
            output.extend(type_lines)
            output.extend(file_lines[insert_line:])

            print(poke_id, t)
            open(f, "w", encoding="utf-8").writelines(output)
